File: Revision/Array/Medium/Next Permutation/Solution1.java
class Solution {
    public void nextPermutation(int[] nums) {
        NextPermutation np = new NextPermutation(nums);
        nums = np.getResult();
    }
}

class NextPermutation{
    private int []  nums;

    NextPermutation(int [] nums){
        this.nums = nums;
    }

    private void findNextPermutation(){
            if(!isValid()){
                return;
            }

            int pivotIndex = -1;

            int len = this.nums.length;

            for(int i=len-2;i>=0;i--){
                if(this.nums[i]<this.nums[i+1]){
                    pivotIndex = i;
                    break;
                }

            }

            if(pivotIndex!=-1){
                int secondPivot = -1;
                for(int i = len-1;i>pivotIndex;i--){
                    if(this.nums[i]>this.nums[pivotIndex]){
                        secondPivot = i;
                        break;
                    }
                }

                if(secondPivot!=-1){
                    this.swap(pivotIndex, secondPivot);
                }
            }

            this.reverse(pivotIndex+1, len-1);
    }

    private void swap(int i, int j){
        int temp = this.nums[i];
        this.nums[i] = this.nums[j];
        this.nums[j] = temp;
    }

    private void reverse(int start, int stop){
        while(start<stop){
            this.swap(start, stop);
            start++;
            stop--;
        }
    }

    private boolean isValid(){
        return this.nums!=null && this.nums.length>0;
    }
    public int[] getResult(){
        this.findNextPermutation();
        return this.nums;
    }
}

File: Revision/Array/Medium/Next Permutation/Solution1.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        n = len(nums)

        pivot_index = -1

        for i in range(n-2, -1, -1):
            if nums[i] < nums[i+1]:
                pivot_index = i
                break
        
        if pivot_index!=-1:
            second_pivot = -1
            for i in range (n-1, pivot_index, -1):
                if nums[i] > nums[pivot_index]:
                    second_pivot = i
                    break
            
            if second_pivot != -1:
                nums[pivot_index], nums[second_pivot] = nums[second_pivot], nums[pivot_index]

        left, right = pivot_index+1, n-1

        while(left < right):
            nums[left], nums[right] = nums[right], nums[left]
            left += 1
            right -= 1

        
